common/src/stack/command/stack/argument_processors/repo.py

```diff
@@ -0,0 +1,178 @@
+from collections import namedtuple
+
+from stack.exception import ArgNotFound
+
+class RepoArgProcessor:
+	REQUIRED_REPO_COLUMNS = [
+		'name',
+		'alias',
+		'url',
+	]
+
+	OPTIONAL_REPO_COLUMNS = [
+		'autorefresh',
+		'assumeyes',
+		'type',
+		'is_mirrorlist',
+		'gpgcheck',
+		'gpgkey',
+		'os',
+		'pallet_id',
+	]
+
+	REPO_COLUMNS = REQUIRED_REPO_COLUMNS + OPTIONAL_REPO_COLUMNS
+
+	def insert_repo(self, name, alias, url, **kwargs):
+		'''
+		Insert a repo with optional kwargs into the repos table.
+		This works like add pallet - if you try to add a repo that already exists, silently succeed
+		'''
+		data = {}
+		for key in self.OPTIONAL_REPO_COLUMNS:
+			if key in kwargs:
+				data[key] = kwargs[key]
+
+		where_clause = ' AND '.join(f'{col}=%s' for col in self.REQUIRED_REPO_COLUMNS + list(data))
+		select_stmt = f'''id FROM repos WHERE {where_clause}'''
+
+		if len(self.db.select(select_stmt, (name, alias, url) + tuple(data.values()))) != 0:
+			return
+
+		cols = ', '.join(['name', 'alias', 'url'] + list(data))
+		vals = ', '.join(['%s'] * (3 + len(data.values())))
+
+		sql = f'''INSERT INTO repos
+			({cols})
+			VALUES
+			({vals})
+			'''
+
+		self.db.execute(sql, (name, alias, url, *data.values()))
+
+	def delete_repo(self, repo):
+		''' delete the repo from the database, should also remove any box associations '''
+		sql = '''DELETE FROM repos WHERE name LIKE BINARY %s OR alias LIKE BINARY %s '''
+		self.db.execute(sql, (repo, repo))
+
+	def enable_repo(self, repo, box):
+		''' add a row to the repo_stacks table, tying a repo to a box '''
+
+		repo_id = self.get_repo_id(repo)
+		box_id = self.db.select('id FROM boxes WHERE name=%s', (box,))[0][0]
+		sql = '''INSERT INTO repo_stacks
+			(box, repo)
+			VALUES
+			(%s, %s)
+			'''
+		self.db.execute(sql, (box_id, repo_id))
+
+	def disable_repo(self, repo, box):
+		''' disable the repo for a given box '''
+		repo_id = self.get_repo_id(repo)
+		box_id = self.db.select('id FROM boxes WHERE name=%s', (box,))[0][0]
+		sql = '''DELETE FROM repo_stacks WHERE box=%s AND repo=%s '''
+		self.db.execute(sql, (box_id, repo_id))
+
+	def get_repo_id(self, repo):
+		rows = self.db.select('''id
+			FROM repos
+			WHERE name LIKE BINARY %s OR alias LIKE BINARY %s
```

Comment:
Not sure you need `BINARY` here, since you have a UNIQUE constraint on the field already.